x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/service/ServiceAccountToken.java

```diff
@@ -0,0 +1,171 @@
+import org.elasticsearch.common.CharArrays;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.hash.MessageDigests;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
+import org.elasticsearch.xpack.core.security.authc.support.UsernamePasswordToken;
+import org.elasticsearch.xpack.core.security.authc.service.ServiceAccount.ServiceAccountId;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+/**
+ * A decoded credential that may be used to authenticate a {@link ServiceAccount}.
+ * It consists of:
+ * <ol>
+ *   <li>A {@link #getAccountId() service account id}</li>
+ *   <li>The {@link #getTokenName() name of the token} to be used</li>
+ *   <li>The {@link #getSecret() secret credential} for that token</li>
+ * </ol>
+ */
+public class ServiceAccountToken implements AuthenticationToken, Closeable {
+
+    public static final String INVALID_TOKEN_NAME_MESSAGE = "service account token name must have at least 1 character " +
+        "and at most 256 characters that are alphanumeric (A-Z, a-z, 0-9) or hyphen (-) or underscore (_). " +
+        "It must not begin with an underscore (_).";
+
+    private static final Pattern VALID_TOKEN_NAME = Pattern.compile("^[a-zA-Z0-9-][a-zA-Z0-9_-]{0,255}$");
+
+    public static final byte MAGIC_BYTE = '\0';
+    public static final byte TOKEN_TYPE = '\1';
+    public static final byte RESERVED_BYTE = '\0';
+    public static final byte FORMAT_VERSION = '\1';
+    public static final byte[] PREFIX = new byte[] { MAGIC_BYTE, TOKEN_TYPE, RESERVED_BYTE, FORMAT_VERSION };
+
+    private static final Logger logger = LogManager.getLogger(ServiceAccountToken.class);
+
+    private final ServiceAccountId accountId;
+    private final String tokenName;
+    private final SecureString secret;
+
+    public ServiceAccountToken(ServiceAccountId accountId, String tokenName, SecureString secret) {
+        this.accountId = accountId;
+        this.tokenName = tokenName;
+        this.secret = secret;
+    }
+
+    public ServiceAccountToken(String qualifiedName, SecureString secret) {
+        final String[] split = Strings.delimitedListToStringArray(qualifiedName, "/");
+        if (split == null || split.length != 3) {
+            throw new IllegalArgumentException(
+                "The qualified name of a service token should take format of 'namespace/service_name/token_name'," +
+                    " got [" + qualifiedName + "]");
+        }
+        this.accountId = new ServiceAccountId(split[0], split[1]);
+        this.tokenName = split[2];
+        this.secret = secret;
```

Comment:
We need null checks on this (or we need to change `equals`)

Reply:
Done